### ISSUE TITLE
Tracking Issue for Path::file_prefix

### ISSUE THREAD

**Issue by mbhall88** [rust-lang/rust#86319]
Feature gate: `#![feature(path_file_prefix)]`

This is a tracking issue for an implementation of the method `std::path::Path::file_prefix`.  It is effectively a "left" variant of the existing [`file_stem`](https://doc.rust-lang.org/std/path/struct.Path.html#method.file_stem) method.

### Public API

```rust
use std::path::Path;

let path = Path::new("foo.tar.gz");
assert_eq!(path.file_prefix(), Some("foo"));
```

### Steps / History

- [ ] Implementation: #85166
- [ ] Final commenting period (FCP)
- [ ] Stabilization PR

### Unresolved Questions

- None yet.

**Comment by schwern** [comment]
How about a corresponding file_suffix? That completes the set. file_stem/extension and file_prefix/file_suffix.

```
    pub fn file_suffix(&self) -> Option<&OsStr> {
        self.file_name().map(split_file_at_dot).and_then(|(_before, after)| Some(after))
    }
```

**Comment by mbhall88** [comment]
@schwern that is indeed a nice idea. I guess it would be best to get the go ahead from someone in the libs team? I don't know how these tracking issues work. The PR for this feature was merged a while ago, but I have no idea if I'm supposed to do something with this issue? @rust-lang/libs

**Comment by cyqsimon** [comment]
Any progress on this? It's a very useful API when working with files like tarballs.

**Comment by mbhall88** [comment]
@cyqsimon it is currently on nightly I believe. Not sure how long it takes to get onto stable though.

As for @schwern's suggestion, I guess that would have to be raised as a separate PR/issue and approved by someone on the libs team.

**Comment by hustcer** [comment]
> How about a corresponding file_suffix? That completes the set. file_stem/extension and file_prefix/file_suffix.

Good point !

**Comment by yescallop** [comment]
One of my concerns is that a `file_prefix` or `file_suffix` method that simply slices the path wouldn't be very much useful, because it is common that users insert dots in the meaningful part of file names, such as `foo.bar.tar.gz`. The current `file_stem` and `extension` methods exist mainly because of File Associations on Windows and common practice on Linux, while it is relatively rare that one would need the exact part of file name before or after the first dot.

I'd suggest adding this to the unresolved questions, if possible.

**Comment by lmapii** [comment]
While it might be impossible to have a 100% correct implementation in my opinion it would already be helpful to simply have all rust crates behave in a known manner. E.g., there are now hundreds of rust crates out there dealing with files, e.g., matching filenames, all behaving a little different. It is impossible to know how a rust crate that you're including will behave.

Even having a method that might change over time and for which it is documented how it behaves in certain corner cases would be very, very useful. Compare it to the `pathlib` in `python`: It is not perfect but it is a beautiful piece of software to use.

**Comment by EvanCarroll** [comment]
> while it is relatively rare that one would need the exact part of file name before or after the first dot.

It's not actually relatively rare. We went over this in voice chat. I gave you multiple reasons why someone would want to do this. In practice, 

* `.tar.gz`
* `.spec.ts`
* `.openapi.yaml`
* `.conf.d` (for an example in a directory)

It's very popular with javascript (where there is no magic byte and everything is js) `.d.ts`, `.conf.json`, `.conf.js`, `.rc.json`, `rc.js`, `.esm.js`, `.cjs.js`. But also with translations `.ru.md` etc. Are all common practices that we have to contend with right now many of these **in the same project**.

----

The [current suggestion to get the functionality is](https://discord.com/channels/442252698964721669/443150878111694848/976154692423917659),

```rust
.file_name().and_then(OsStr::to_str).map_or(false, |name| name.ends_with(".tar.gz"))
```

Which I think is a bit convoluted. I am down for doing this other ways, but I don't think something like this should is "rare" as far as path manipulation goes, or that it should require converting to-and-from `str`. I think the simple-most question here is 

* _Do we want `file_suffix` to return the `.` that `file_prefix` leaves?_ My own preference, I think `file = file_prefix + file_suffix` if we go down that route.

We could potentially find better ways to do all of this though. Like simply returning an iterator for `file_parts` where given `foo.bar.tar.gz` could be,

* `.next().collect::<OsStr>()`, what I think of as extension (`.bar.tar.gz`), and I'm patching to be called `.file_suffix()` (tail)
* `.last()`, what we currently call `.extension()`
* `.first()` what we currently call `.file_prefix()` (head)
* `file_stem` would be all but the last element (not sure if there is a nice way to do that in Rust).. (init)

The bottom line here is that when it comes to the file name, people will logically think of splitting on `.` and want names for all possible sets/strings of that list, (from the List Monster in Learn You a Haskell)

![zj2Lb](https://user-images.githubusercontent.com/19922/168976953-a1e4e488-21d6-4dd5-9078-d5b543d804f7.png)

Something about that API of returning an Iterator just seems simpler than arguing about whether an "extension" in `foo.tar.gz` is `.tar.gz`, `tar.gz`, `.gz` or `gz`.

**Comment by yescallop** [comment]
There might've been some misunderstandings here. Let's take this file name `File Version 3.5.4 (Copy).tar.gz` (from Jacob Lifshay) as an example.
In this case, `.file_prefix()` returns `File Version 3` while `.file_suffix()` returns `.5.4 (Copy).tar.gz`. If you need to determine if this file name ends with `.tar.gz` or not, you still need to do stuffs like
```rust
.and_then(OsStr::to_str).map_or(false, |name| name.ends_with(".tar.gz"))
```
IMO this is not optimal. Maybe we could add methods `ends_with` and `strip_suffix` on `OsStr` if this is sufficient for most use cases. But things are getting too complicated in here.

**Comment by mbhall88** [comment]
> We could potentially find better ways to do all of this though. Like simply returning an iterator for file_parts where given foo.bar.tar.gz 

I think this is probably the "correct" way to approach this. With such an iterator method we can easily implement any/all flavours of splitting up file names.

**Comment by yescallop** [comment]
FYI, here's [my attempt](https://github.com/yescallop/rust/commit/1a36e5147c2f32cafa8a7d5e5596a823789e97a4) to implement `Path::extensions` which is akin to `suffixes` in python. Your comments are very welcome.
```rust
let mut exts = Path::new("foo.tar.gz").extensions().unwrap();
assert_eq!(exts.next(), Some(OsStr::new("gz")));
assert_eq!(exts.next(), Some(OsStr::new("tar")));
assert_eq!(exts.next(), None);
assert_eq!(exts.visited(), "tar.gz");
assert_eq!(exts.remaining_stem(), "foo");
```

**Comment by yescallop** [comment]
Correct me if I'm wrong, but I believe that having a method that extracts the part of file name before or after the **first** dot is not a good idea. By splitting a file name at the **first** dot, one asserts that they won't have dots in the meaningful part of file name. However, there are actually a bunch of densely-dotted file names in the wild that go against this design, for example:

`pyroute2.nslink-0.6.9.tar.gz`, `api.admin.users.service.spec.ts`, `Demo.Sales.APISvc.openapi.yaml`, `sap.ui.webc.common.d.ts`, `proxy.remote.docker.eximee.conf.json`, `webpack.dist.components.px.conf.js`, `org.freebsd.rc.json`, and so on.

IMO, none of these file names benefits much from a split at the **first** dot unless it is guaranteed that we deny dots in a meaningful name. I'd prefer to go with the iterator solution suggested above and I'm willing to open a PR for this.

**Comment by EvanCarroll** [comment]
>  I believe that having a method that extracts the part of file name before or after the first dot is not a good idea. By splitting a file name at the first dot, one asserts that they won't have dots in the non-extension part of file name.

Agreed, that's why I named the original suggestion `file_parts` and not `extension`.

> We could potentially find better ways to do all of this though. Like simply **returning an iterator for `file_parts`** where given foo.bar.tar.gz could be ([list monster](http://learnyouahaskell.com/starting-out) in graphic below),

**Comment by yescallop** [comment]
@EvanCarroll That might be a solution, but I think there are some disadvantages to it:

The `file_parts` method returns a *forward* iterator over the parts of file name. I think this is against the fact that it is often the case that a program considers the non-extension part of file name **opaque** to path processing. As long as the extensions match, the program may extract the remaining part of file name and pass it to another function for further parsing, whether there are dots in it or not.

For this reason, a *forward* iterator seems not very practical. Also, I'm afraid it is not allowed to do a `.collect::<&OsStr>()` on an iterator that yields `OsStr` slices, because that requires `&OsStr: FromIterator<&OsStr>`. Doing a `.collect::<OsString>()` is possible, but it does unnecessary allocation.

This is why I implemented `Path::extensions` and two methods on the resulting `Extensions` struct:
- `.visited()` to return the visited extensions as a consecutive `OsStr`.
- `.remaining()` to return the part of file name remaining for iteration.

If it is really needed to do what `.file_prefix()` does right now, we may have:
- `.skip_all()` to skip all the remaining extensions. Then you call `.remaining()` to get `file_prefix` or `.visited()` to get `file_suffix`.

And if it is sometimes useful to return the extensions with preceding dots, we could have:
- `.with_preceding_dots()` to make the iterator do that.

**Comment by mbhall88** [comment]
> Correct me if I'm wrong, but I believe that having a method that extracts the part of file name before or after the **first** dot is not a good idea. By splitting a file name at the **first** dot, one asserts that they won't have dots in the non-extension part of file name. However, there are actually a bunch of densely-dotted file names in the wild that go against this design, for example:
> 
> `pyroute2.nslink-0.6.9.tar.gz`, `api.admin.users.service.spec.ts`, `Demo.Sales.APISvc.openapi.yaml`, `sap.ui.webc.common.d.ts`, `proxy.remote.docker.eximee.conf.json`, `webpack.dist.components.px.conf.js`, `org.freebsd.rc.json`, and so on.
> 
> IMO, none of these file names benefits much from a split at the **first** dot unless it is guaranteed that we deny dots in the non-extension part of a file name. I'd prefer to go with the iterator solution suggested above and I'm willing to open a PR for this.

You are right, none of those listed file names benefit from a split at the first dot. But there are **many more** file names "in the wild" that do. Solution: anyone working with those file names uses some other method to extract the parts they want...  
Saying a method is not needed because some corner cases make it useless doesn't negate it being very useful in a whole bunch of other circumstances.

**Comment by robmv** [comment]
Instead of creating new methods with new semantics that will not treat files with dots on parts of the filename not related to the extension, why not extend the current ones like:

```rust
let path = Path::new("foo 1.2.3.tar.gz");
assert_eq!(path.file_stem_multi(2), Some("foo 1.2.3"));
assert_eq!(path.extension_multi(2), Some("tar.gz"));
```

This will only have a different contract than the current ones, where:

    Otherwise, the portion of the file name before the final `.`

is changed to;

    Otherwise, the portion of the file name before the nth `.` from the end.

Where `file_stem()` and `extension()` are just a simple case of these new methods with the argument `1`. Another helper methods like `with_extension(..)` could be extended to indicate the number of segments of the extension wanted to be replaced.

Note: named with s better suffix than `_multi`.

**Comment by mbhall88** [comment]
@thomcc apologies for pinging you directly but I couldn't find a tag for the library team anywhere. `@dtolnay` was originally managing this feature but I don't know if he is on the library team anymore? I don't really know how long these features normally take to progress through to stable but this issue has been open for two and a half years which seems like longer than I would expect?  
Can you suggest any next steps to get this stabilised?

**Comment by tgross35** [comment]
> @thomcc apologies for pinging you directly but I couldn't find a tag for the library team anywhere. `@dtolnay` was originally managing this feature but I don't know if he is on the library team anymore? I don't really know how long these features normally take to progress through to stable but this issue has been open for two and a half years which seems like longer than I would expect? Can you suggest any next steps to get this stabilised?

Opening a stabilization PR is trivial, you just need to change the `unstable` gate to `stable` as described here https://rustc-dev-guide.rust-lang.org/stability.html#stabilizing-a-library-feature.

That being said, I have to agree with the concerns that the algorithm here can be a bit bikeshed-y and this could be better served by pattern functions on `OsStr`, which has been proposed https://github.com/rust-lang/libs-team/issues/311.

Opening a stabilization PR to start libs-api discussion probably isn't a bad idea in any case.

**Comment by mbhall88** [comment]
@tgross35 according to the stabilization docs you linked the first step is 

> Ask a **@T-libs-api** member to start an FCP on the tracking issue and wait for the FCP to complete (with `disposition-merge`).

Are you a member of said team? If I try and use that identifier in this comment it doesn't come up with anything...

**Comment by tgross35** [comment]
I am not, but team members can always be found in the relevant files on the teams repo https://github.com/rust-lang/team/blob/master/teams/libs-api.toml. Pinging team members isn't really the best way to get something on the radar though, so opening the stabilization PR first and requesting libs-api review is usually easier for trivial things like this (FCP just happens there).

If you have an idea for an iterator-based API it may be worth proposing that too as an alternative. New API proposals go through the ACP process, which is an issue template at https://github.com/rust-lang/libs-team/issues.

**Comment by mbhall88** [comment]
Thanks very much for you help @tgross35. I'll open the PR and we can discuss the FCP there and whether this implementation or an iterator API is preferable.

**Comment by joshtriplett** [comment]
Shall we stabilize this?

@rfcbot merge

**Comment by the8472** [comment]
Concern: If we're already doing the work to parse the suffix we might as well return both a suffix and prefix, since they're just borrows.

Possible names:

* `name_and_extensions()`
* `stem_and_suffix()`
* `name_affixes()`
* `name_parts()`

**Comment by mbhall88** [comment]
Do you mean adding another function with that functionality? 

Because the function being tracked here, `Path::file_prefix` is essentially a counterpart to [`Path::file_stem`](https://doc.rust-lang.org/std/path/struct.Path.html#method.file_stem), which only returns the prefix - so I was trying to keep the API consistent in that sense.

I do agree that having such function(s) would be useful though.

**Comment by the8472** [comment]
> Because the function being tracked here, Path::file_prefix is essentially a counterpart to [Path::file_stem](https://doc.rust-lang.org/std/path/struct.Path.html#method.file_stem), which only returns the prefix - so I was trying to keep the API consistent in that sense.

I suppose it depends on what one uses them for. For `foo.txt` they both do the same. But for `foo.txt.gz` they do something quite different. `file_prefix` provides the base name (good for naming things) while `file_stem` only removes one layer of format nesting (good for processing things layer by layer).

So the question is what one does with this method. If you need to know the full extension (e.g. `.tar.gz` or `.conf.d`) to do further categorization then we currently have no way to extract that. And of course we could add yet another method do that, but then you'd have to parse twice - or do some slicing based on offsets which is a bit more to write.
So my though was that if we're adding a new API then we might as well parse once and return both values.

**Comment by mbhall88** [comment]
Should I close this issue or do something to remove it from nightly? It's been over 4 years now and it feels like there has been a lot of push back so maybe it's just easier for everyone to move on and forget about it?

**Comment by kaleidawave** [comment]
I agree this should be closed. It is not well specified of paths with *2 or more extensions*.

It is relatively simple to implement this for your use case. For example in the following, the predicate `chr.is_ascii_whitespace() || matches!(chr, b'/' | b'\\' | b'-' | b'_')` can be modified based on what is deemed not a valid extension

```rust
/// modified source from https://doc.rust-lang.org/stable/src/std/path.rs.html#2721-2723
mod utilities {
    use std::ffi::OsStr;

    pub fn file_suffix(this: &std::path::Path) -> Option<&OsStr> {
        this.file_name()
            .map(split_file_at_dot)
            .and_then(|(_before, after)| after)
    }

    fn split_file_at_dot(file: &OsStr) -> (&OsStr, Option<&OsStr>) {
        if file.as_encoded_bytes() == b".." {
            return (file, None);
        }

        let slice = file.as_encoded_bytes();
        let mut ridx = None;
        for (idx, chr) in slice.iter().enumerate().rev() {
            if let b'.' = chr {
                ridx = Some(idx);
            } else if chr.is_ascii_whitespace() || matches!(chr, b'/' | b'\\' | b'-' | b'_') {
                break;
            }
        }
        if let Some(ridx) = ridx {
            let (left, right) = (&slice[..ridx], &slice[ridx + 1..]);

            // The unsafety here stems from converting between &OsStr and &[u8]
            // and back. This is safe to do because (1) we only look at ASCII
            // contents of the encoding and (2) new &OsStr values are produced
            // only from ASCII-bounded slices of existing &OsStr values.
            unsafe {
                (
                    OsStr::from_encoded_bytes_unchecked(left),
                    Some(OsStr::from_encoded_bytes_unchecked(right)),
                )
            }
        } else {
            (file, None)
        }
    }
}
```
